module/os/linux/spl/spl-taskq.c

```diff
@@ -848,12 +854,35 @@ taskq_thread_should_stop(taskq_t *tq, taskq_thread_t *tqt)
 	    tqt_thread_list) == tqt)
 		return (0);
 
-	return
+	int no_work =
 	    ((tq->tq_nspawn == 0) &&	/* No threads are being spawned */
 	    (tq->tq_nactive == 0) &&	/* No threads are handling tasks */
 	    (tq->tq_nthreads > 1) &&	/* More than 1 thread is running */
 	    (!taskq_next_ent(tq)) &&	/* There are no pending tasks */
 	    (spl_taskq_thread_dynamic)); /* Dynamic taskqs are allowed */
+
+	/*
+	 * If we would have said stop before, let's instead wait a bit, maybe
+	 * we'll see more work come our way soon...
+	 */
+	if (no_work) {
+		/* if it's 0, we want the old behavior. */
+		if (spl_taskq_thread_timeout_ms == 0)
+			return (1);
```

Comment:
Let's also return here when `!(tq->tq_flags & TASKQ_ACTIVE)`, this is set at the top of `taskq_destroy()`.

Reply:
return 0 or 1, though? I would assume 1, in that case, but I'm not sure I understand the failure mode - is it that this might not always say yes to destroying when tearing down the taskq, so we might end up with living threads without their taskq because we assume that can't happen?

Reply:
> we might end up with living threads without their taskq

That's my guess.  Even though the logic [here](https://github.com/openzfs/zfs/blob/master/module/os/linux/spl/spl-taskq.c#L1189-L1205) should block waiting for all the threads to exit.  I'd have thought if we missed a wake the symptom would be `taskq_destroy()` never exits.  It's more like `kthread_stop()` returned before the thread exited.

We can count on no new tasks being added once `taskq_destroy()` is called, clearing `TASKQ_ACTIVE` will prevent this.  After which `task_wait()` is invoked to wait for everything outstanding to complete.  At that point we want to make sure `taskq_thread_should_stop()` returns 1 so we don't get stuck blocking in `taskq_destroy()` longer than we need too.